mempool/src/tests/test_framework.rs

```diff
@@ -0,0 +1,516 @@
+        .all(|txn| block_contains_test_transaction(block, txn))
+        && block.len() == txns.len()
+}
+
+pub fn block_contains_any_test_transactions(
+    block: &[SignedTransaction],
+    txns: &[TestTransaction],
+) -> bool {
+    txns.iter()
+        .any(|txn| block_contains_test_transaction(block, txn))
+        && block.len() == txns.len()
+}
+
+fn block_contains_test_transaction(block: &[SignedTransaction], txn: &TestTransaction) -> bool {
+    block.iter().any(|signed_txn| {
+        signed_txn.sequence_number() == txn.sequence_number
+            && signed_txn.sender() == TestTransaction::get_address(txn.address)
+    })
+}
+
+pub async fn send_message(
+    node: &mut MempoolNode,
+    protocol_id: ProtocolId,
+    network_id: NetworkId,
+    remote_peer_id: PeerId,
+    txns: &[TestTransaction],
+) {
+    let inbound_handle = node.node_inbound_handle(network_id);
+    let request_id = Vec::new();
+    let msg = MempoolSyncMsg::BroadcastTransactionsRequest {
+        request_id: request_id.clone(),
+        transactions: sign_transactions(txns),
+    };
+    let data = protocol_id.to_bytes(&msg).unwrap().into();
+    let (notif, maybe_receiver) = match protocol_id {
+        ProtocolId::MempoolDirectSend => (
+            PeerManagerNotification::RecvMessage(
+                remote_peer_id,
+                Message {
+                    protocol_id,
+                    mdata: data,
+                },
+            ),
+            None,
+        ),
+        ProtocolId::MempoolRpc => {
+            let (res_tx, res_rx) = oneshot::channel();
+            let notif = PeerManagerNotification::RecvRpc(
+                remote_peer_id,
+                InboundRpcRequest {
+                    protocol_id,
+                    data,
+                    res_tx,
+                },
+            );
+            (notif, Some(res_rx))
+        }
+        _ => panic!("Invalid protocol"),
+    };
+    inbound_handle
+        .inbound_message_sender
+        .push((remote_peer_id, protocol_id), notif)
+        .unwrap();
+
+    let response: MempoolSyncMsg = if let Some(res_rx) = maybe_receiver {
+        let response = res_rx.await.unwrap().unwrap();
+        protocol_id.from_bytes(&response).unwrap()
+    } else {
+        match node.outbound_handle(network_id).next().await.unwrap() {
+            PeerManagerRequest::SendDirectSend(peer_id, msg) => {
+                assert_eq!(peer_id, remote_peer_id);
+                msg.protocol_id.from_bytes(&msg.mdata).unwrap()
+            }
+            _ => panic!("Should not be getting an RPC response"),
+        }
+    };
+    if let MempoolSyncMsg::BroadcastTransactionsResponse {
+        request_id: response_request_id,
+        retry,
+        backoff,
+    } = response
+    {
+        assert_eq!(response_request_id, request_id);
+        assert!(!retry);
+        assert!(!backoff);
+    } else {
+        panic!("Expected a response!");
+    }
+}
+
+pub async fn respond_to_request(
```

Comment:
nit: maybe call this `verify_response_sent`? I mean, technically it's asserting that the node generated a response and has sent it along the wire?

Reply:
Well, it asserts the response is sent, and returns an ack.  Which is I guess why I've got the "respond to"

Reply:
How about `verify_broadcast_and_ack`?